### ISSUE TITLE
Prepared Statement and NVARCHAR(MAX)

### ISSUE THREAD

**Issue by psevdaisthisi** [TimelordUK/node-sqlserver-v8#248]
Hi,

I'm having an issue querying fields of the type `NVARCHAR(MAX)` using a Prepared Statement.
It looks like all characters in the field are replaced by `\x00` control sequences in the result.

Consider the following test.
```sql
const sql = require("msnodesqlv8");

const hostname = "...";
const username = "...";
const password = "...";
const driver = "ODBC Driver 18 for SQL Server";
const conn_str = `Server=${hostname};UID=${username};PWD=${password};Encrypt=No;TrustServerCertificate=No;Driver=${driver}`;

sql.open({ conn_str }, function (err, conn) {
	const q =" DECLARE @v NVARCHAR(MAX) = 'hello'; SELECT @v AS v";
	conn.prepare(q, function (err, ps) {
		ps.preparedQuery([], function (err, out) {
			process.stdout.write ("Prepared Statement: ");
			console.log(out[0]);
			ps.free();
		});
	});
});

sql.open({ conn_str }, function (err, conn) {
	const q =" DECLARE @v NVARCHAR(MAX) = 'hello'; SELECT @v AS v";
	conn.query(q, function (err, out) {
		process.stdout.write ("Query: ");
		console.log(out[0]);
	});
});
```

When running it, I get the following result:
```
Query: { v: 'hello' }
Prepared Statement: { v: '\x00\x00\x00\x00\x00' }
```
So it works fine under a "regular" query but not under a Prepared Statement.
At first I though it was because I was using `ODBC Driver 18 for SQL Server` but I got the same exact result under v17. Reproduced this under Linux and Windows using `msnodesqlv8` version `2.4.8` and SQL Server versions `14.0.3436.1` and `15.0.2080.9`.

Any idea what might be the problem? If I replace `NVARCHAR(MAX)` by something like `NVARCHAR(64)` it works fine. Just note that this is a toy example to make it easy for others to try it. In reality I got this issues when running Prepared Statements against tables that have `NVARCHAR(MAX)` columns.

Thank you!

**Comment by TimelordUK** [comment]
this is now on master but not yet released- basically you have to constrain this column as rows are allocated into memory fixed width for prepared statements . The driver now allocates a dfault of 8k for nvarchar max on prepared queries - out box your example would work else the column would be truncated.

you can also provide a hint as below which tells driver to use this size - keep in  mind if you have many columns 50 rows worth are allocared so the memory will quickly grow depending on size.

for non prepared queries we keep pulling the column rather like a garden bird pulling a worm out ground until it is finished - hence for non prepared we dont see this issue.

it is stil in test and a couple of things are left outstanding before release.


```js
  it('use prepared and select nvarchar max with max default size on connection', async function handler () {
    const s = 'hello'
    const max = 4
    const q = `DECLARE @v NVARCHAR(MAX) = '${s}'; SELECT @v AS v`
    env.theConnection.setMaxPreparedColumnSize(max)
    const prepared = await env.theConnection.promises.prepare(q)
    const res = await prepared.promises.query([])
    assert.deepStrictEqual(res.first[0].v, s.slice(0, 4))
    await prepared.free()
  })

  it('use prepared and select nvarchar max with max default size on query', async function handler () {
    const s = 'hello'
    const max = 4
    const q = {
      query_str: `DECLARE @v NVARCHAR(MAX) = '${s}'; SELECT @v AS v`,
      max_prepared_column_size: max
    }
    const prepared = await env.theConnection.promises.prepare(q)
    const res = await prepared.promises.query([])
    assert.deepStrictEqual(res.first[0].v, s.slice(0, 4))
    await prepared.free()
  })

  it('use prepared and select nvarchar(max)', async function handler () {
    const s = 'hello'
    const q = `DECLARE @v NVARCHAR(MAX) = '${s}'; SELECT @v AS v`
    const prepared = await env.theConnection.promises.prepare(q)
    const res = await prepared.promises.query([])
    assert.deepStrictEqual(res.first[0].v, s)
    await prepared.free()
  })

```

**Comment by psevdaisthisi** [comment]
Thank you so much for putting the time into solving this issue and for the heads up about memory consumption tradeoffs.  I've cloned the repo a few minutes ago, built the driver under Linux/x64 and tried it on my project. Worked flawlessly with both the default size and with `max_prepared_column_size`.

**Comment by psevdaisthisi** [comment]
Closing because this was fixed in [v2.5.0](https://github.com/TimelordUK/node-sqlserver-v8/releases/tag/v2.5.0).
Once again, thank you.
